src/p2lab/genetic/matching.py

```diff
@@ -0,0 +1,44 @@
+import networkx as nx
+import numpy as np
+
+def dense(teams : list[object]) -> np.ndarray:
+	r"""
+	Generates a fully connected set of edges. Each team is 
+	connected to every other one.
+
+	Inputs:
+		:teams: set of teams.
+
+	Outputs:
+	
+		np.ndarray shape (N_edges, 2)
+		
+	"""
+	g = nx.complete_graph(len(teams))
+
+	# Select only the node index pairs, exclude the feature dict.
+	return np.stack( g.edges(data=False) )
+
+def mst(teams : list[object]) -> np.ndarray:
+	r"""
+	Returns a minimum spanning tree graph generated from the
+	fully connected graph over the teams.
```

Comment:
Does this return an absolute minimum? It is possible to sample > minimum, < complete graph, such that everying is connected but we may have more connections that minimally required?

Reply:
Hmm, I think by default the graph mst runs over has all edges weighted equally, so min spanning tree should be the min number of edges so that all nodes are reachable. I am not sure that you can drop an edge and still have no disconnected components?

Reply:
I mean increasing rather than dropping edges - so if M is the minimum number of edges and N is the max, is it possible to keep additional edges above the minimum such that the total number of edges T is

M < T < N

if not, don't worry - I suspect this is nice to have instead of essential!

Reply:
Yeah, we could add it as a feature. One thing I've noticed is that mst creates the spanning tree by simply linking one node to all the rest, so you get a fairly trivial spanning tree.

Reply:
Okay - I leave it up to you whether or not to implement this now or to leave it as a potential addition later. If later, this looks ready to merge to me!

One thought is it could be parameterised as a value between 0 and 1, where 0 means we have M edges, 1 means we have N, 0.5 T = (N+M)/2 edges, etc